File: src/AoC16/s16e01-rust/part2.rs
use std::io::{self, Read};

mod common;
use common::solve_part2;

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let result = solve_part2(&input);
    println!("{}", result);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sample_revisit() {
        // R8, R4, R4, R8 - first location visited twice is 4 blocks away, due East
        let input = "R8, R4, R4, R8";
        assert_eq!(solve_part2(input), 4);
    }
}

File: src/AoC16/s16e01-zig/part2.zig
const std = @import("std");
const common = @import("common.zig");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdin = std.io.getStdIn().reader();
    const input = try stdin.readAllAlloc(allocator, 1024 * 1024);
    defer allocator.free(input);

    const result = try common.solvePart2(allocator, std.mem.trim(u8, input, "\n"));
    const stdout = std.io.getStdOut().writer();
    try stdout.print("{d}\n", .{result});
}

test "part2_sample_1" {
    // R8, R4, R4, R8 - first location visited twice is 4 blocks away, due East
    const result = try common.solvePart2(std.testing.allocator, "R8, R4, R4, R8");
    try std.testing.expectEqual(@as(i32, 4), result);
}
